### ISSUE TITLE
GANs

### ISSUE THREAD

**Issue by YunchaoYang** [YunchaoYang/Blogs#29]
# Ways to improve GAN performance
3 major issues 
1. non-convergence
2. mode collapse
3. slow training

Ways
1. cost function for better optimization goal
2. add additional penalties to enforce constraints
3. avoid overconfidence and overfitting
4. better ways of optimizing modesl
5. add labels

`The generator tries to find the best image to fool the discriminator. The “best” image keeps changing when both networks counteract their opponent. However, the optimization can turn too greedy and fall it into a never-ending cat-and-mouse game. ` - Jonathan Hui

- Feature matching expands the goal from beating the opponent to matching features in real images. 

When mode collapses, all images created looks similar. To mitigate the problem, we feed real images and generated images into the discriminator separately in different batches and compute the similarity of the image x with images in the same batch. We append the similarity o(x) in one of the dense layers in the discriminator to classify whether this image is real or generated.

- Minibatch discrimination

`Deep networks may suffer from overconfidence. For example, it uses very few features to classify an object. To mitigate the problem, deep learning uses regulation and dropout to avoid overconfidence.`
In GAN, if the discriminator depends on a small set of features to detect real images, the generator may just produce these features only to exploit the discriminator. The optimization may turn too greedy and produces no long term benefit. In GAN, overconfidence hurts badly. To avoid the problem, we penalize the discriminator when the prediction for any real images go beyond 0.9 (D(real image)>0.9). This is done by setting our target label value to be `0.9` instead of 1.0.

- One-sided label smoothing

References
1. https://towardsdatascience.com/gan-ways-to-improve-gan-performance-acf37f9f59b
